### DISCUSSION TITLE
Request Reply replace REST

### DISCUSSION THREAD

**Discussion by alberk8** [nats-io/nats.net#443] [Q&A]
Is a good approach to use request reply to replace REST API for micro services between aspnet app?

**Comment by mtmk** [comment]
You can use the Services API as well https://nats-io.github.io/nats.net.v2/documentation/services/intro.html

edit: Coincidentally Jeremy just released a video on YouTube about micro services which might be useful.

**Reply by alberk8** [reply]
Thank you and I have watched the video.  Giving me lots of idea.    
Another question,  Where is the best place to place the Services API in an Aspnet project,  should it be in HostedService or is there any other way like `app.MapController()`  eg `app.MapNatsService<T>()`?

**Reply by mtmk** [reply]
I think hosted service is the easiest at the moment. it is a developing area though. feel free to suggest other implementations.

* edit: you might be interested @adamreed90 just opened #444

**Reply by iamadamreed** [reply]
I'm still flushing out a method for using attributes to register service endpoints, to provide a workflow more in line with traditional api controllers.

Take a look and let me know what you think @alberk8 👍

Here is the updated one, the last one was closed:
https://github.com/nats-io/nats.net.v2/pull/445

**Reply by alberk8** [reply]
I looks good but when I read the code the first thing I wonder is  where does the class `MyClass` get initialized?

I personally prefer the `app.MapNatService<T>()`  if that is possible,  it is so much easier to track down (follow the code).

**Reply by iamadamreed** [reply]
```
   var registrar = INatsSvcEndpointRegistrar.GetRegistrar();
   await registrar.RegisterEndpointsAsync(testService);
   ```
   
   The method `RegisterEndpointsAsync` uses Source Generation to scan the entire project for methods tagged with 
   ` [ServiceEndpoint("endpointName", "queueGroup")]`  and registers them using `AddEndpointAsync<T>`
   
   I am working on another flavor of this where it mimics the functionality of `ControllerBase` to work similarly to an API Controller in ASP.NET

**Reply by alberk8** [reply]
Thank you,  This going to be awesome. 👍

**Reply by alberk8** [reply]
@adamreed90 ,  I cloned the repo with your pull request,  Running the EndpointRegistrationTest,  
The line hitting the error is  `await registrar.RegisterEndpointsAsync(testService)`. 
```
Exception thrown: 'System.NullReferenceException' in EndpointRegistrationTest.dll
Object reference not set to an instance of an object.
```

**Reply by iamadamreed** [reply]
Not sure why you're getting that error, but this was just a proof of concept, I've started a new approach that I'll be finishing up this weekend most likely. Stay tuned!
